tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -6504,6 +6550,11 @@ bool HLSLExternalSource::MatchArguments(
              "have as many arguments and types as the intrinsic template");
   }
 
+  // For object return types that need to match arguments, we need to slot in the full type here
+  // Can't do it sooner because when return is encountered above, the other arg types haven't been set
+  if (Template[pIntrinsic->pArgs[0].uTemplateId] == AR_TOBJ_OBJECT)
+    argTypes[0] = argTypes[pIntrinsic->pArgs[0].uComponentTypeId];
```

Comment:
Man, this function's code is scary/tricky.  I thought this shouldn't be necessary at first, but it seems like this is necessary specifically for object types in this case, since it would normally have constructed a type if it was a basic type (which you can't do for the specific object), or taken it from `objectElement` or `functionTemplateTypeArg`.  This is yet another case.